Modules/core/Controller/CoretilesController.php

```diff
@@ -99,4 +107,99 @@ public function docAction(){
         ));
     }
 
+
+    /**
+     * Distinctly list spaces:
+     * - of which user is member
+     * - in which user has a pending request to join
+     * - of which user is admin
+     * 
+     * @return array of arrays: [userSpaceIds, userPendingSpaceIds, SpacesUserIsAdminOf]
+     */
+    public function getUserSpaces() {
+        $modelSpacePending = new CorePendingAccount();
+        $data = $modelSpacePending->getSpaceIdsForPending($_SESSION["id_user"]);
+        $userPendingSpaceIds = array();
+
+        if ($data && count($data) > 0) {
+            foreach ($data as $space) {
+                array_push($userPendingSpaceIds, $space["id_space"]);
+            }
+        }
+
+        $modelSpaceUser = new CoreSpaceUser();
+        $data = $modelSpaceUser->getUserSpaceInfo($_SESSION["id_user"]);
+        $userSpaceIds = array();
+        $spacesUserIsAdminOf = array();
+
+        if ($data && count($data) > 0) {
+            foreach ($data as $space) {
+                array_push($userSpaceIds, $space["id_space"]);
+                if ($space["status"] === "4") {
+                    array_push($spacesUserIsAdminOf, $space["id_space"]);
+                }
+            }
+        }
+        $result = array(
+            "userSpaceIds" => $userSpaceIds,
+            "userPendingSpaceIds" => $userPendingSpaceIds,
+            "spacesUserIsAdminOf" => $spacesUserIsAdminOf
+        );
+
+        return $result;
+    }
+
+    /**
+     * 
+     * Manage actions resulting from user request to join or leave a space
+     *
+     * @param int $id_space
+     * @param bool $isMemberOfSpace
+     */
+    public function selfJoinSpaceAction($id_space) {
+        $modelSpaceUser = new CoreSpaceUser();
+        $isMemberOfSpace = $modelSpaceUser->exists($_SESSION["id_user"], $id_space);
+        if ($isMemberOfSpace) {
+            $modelSpaceUser = new CoreSpaceUser();
+            $modelSpaceUser->delete($_SESSION["id_user"], $id_space);
+        } else if (!$isMemberOfSpace) {
+            $spaceModel = new CoreSpace();
+            $spaceName = $spaceModel->getSpaceName($id_space);
+            $modelSpacePending = new CorePendingAccount();
+            $modelSpacePending->add($_SESSION["id_user"], $id_space);
```

Comment:
you do not check here if user is already in pending

if already in pending and a new join request is done, you will have multiple entries for same user/space in db